File: tools/calendar/reminders.py
#!/usr/bin/env python3
"""Telegram calendar reminders generator.

- Reads local vdirsyncer .ics files (all calendars under ~/.local/share/vdirsyncer/calendars/)
- Computes reminders in America/New_York
- De-dupes via a local state file

This script does NOT send messages; it prints reminders to stdout.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from datetime import datetime, timedelta, time
from pathlib import Path
from typing import Iterable, List

import pytz
from icalendar import Calendar
import recurring_ical_events

ET = pytz.timezone("America/New_York")

CAL_ROOT = Path(os.path.expanduser("~/.local/share/vdirsyncer/calendars"))
STATE_PATH = Path(os.path.expanduser("~/.local/share/clawdbot/calendar-reminders-state.json"))

IMPORTANT_KEYWORDS = [
    "IMPORTANT",
    "TRAVEL",
    "FLIGHT",
    "DEADLINE",
    "INTERVIEW",
    "APPT",
]


@dataclass(frozen=True)
class Occ:
    cal_name: str
    uid: str
    title: str
    start: datetime
    end: datetime
    all_day: bool


def _load_state() -> dict:
    if STATE_PATH.exists():
        try:
            return json.loads(STATE_PATH.read_text())
        except Exception:
            return {}
    return {}


def _save_state(state: dict) -> None:
    STATE_PATH.parent.mkdir(parents=True, exist_ok=True)
    tmp = STATE_PATH.with_suffix(".tmp")
    tmp.write_text(json.dumps(state, indent=2, sort_keys=True))
    tmp.replace(STATE_PATH)


def _to_et(dt: datetime) -> datetime:
    if dt.tzinfo is None:
        return ET.localize(dt)
    return dt.astimezone(ET)


def _event_title(event) -> str:
    try:
        return str(event.get("summary") or "(no title)")
    except Exception:
        return "(no title)"


def _event_uid(event) -> str:
    try:
        return str(event.get("uid") or "")
    except Exception:
        return ""


def _is_all_day(start: datetime, end: datetime) -> bool:
    s = _to_et(start)
    e = _to_et(end)
    return s.time() == time(0, 0) and (e - s) >= timedelta(days=1) and e.time() == time(0, 0)


def _iter_ics_bytes() -> Iterable[tuple[str, bytes]]:
    if not CAL_ROOT.exists():
        return
    for cal_dir in sorted([p for p in CAL_ROOT.iterdir() if p.is_dir()]):
        cal_name = cal_dir.name
        for p in sorted(cal_dir.glob("*.ics")):
            try:
                yield cal_name, p.read_bytes()
            except Exception:
                continue


def _iter_occurrences(now: datetime, horizon: datetime) -> Iterable[Occ]:
    for cal_name, blob in _iter_ics_bytes():
        try:
            cal = Calendar.from_ical(blob)
        except Exception:
            continue

        occ_events = recurring_ical_events.of(cal).between(now, horizon)
        for ev in occ_events:
            try:
                uid = _event_uid(ev)
                title = _event_title(ev)

                dtstart = ev.decoded("dtstart")
                dtend = ev.decoded("dtend") if ev.get("dtend") else None

                # date-only events come as datetime.date
                if hasattr(dtstart, "year") and not isinstance(dtstart, datetime):
                    start_dt = ET.localize(datetime(dtstart.year, dtstart.month, dtstart.day, 0, 0))
                    if dtend is not None and not isinstance(dtend, datetime):
                        end_dt = ET.localize(datetime(dtend.year, dtend.month, dtend.day, 0, 0))
                    else:
                        end_dt = start_dt + timedelta(days=1)
                    all_day = True
                else:
                    start_dt = _to_et(dtstart)
                    if dtend is None:
                        end_dt = start_dt + timedelta(hours=1)
                    else:
                        if hasattr(dtend, "year") and not isinstance(dtend, datetime):
                            end_dt = ET.localize(datetime(dtend.year, dtend.month, dtend.day, 0, 0))
                        else:
                            end_dt = _to_et(dtend)
                    all_day = _is_all_day(start_dt, end_dt)

                yield Occ(cal_name=cal_name, uid=uid, title=title, start=start_dt, end=end_dt, all_day=all_day)
            except Exception:
                continue


def _reminder_offsets(occ: Occ) -> List[timedelta]:
    title_u = occ.title.upper()
    important = any(k in title_u for k in IMPORTANT_KEYWORDS)

    if occ.all_day:
        return [timedelta(hours=0)]  # special-cased

    dur = occ.end - occ.start

    if important:
        return [timedelta(days=1), timedelta(hours=2), timedelta(minutes=30), timedelta(minutes=10)]

    if dur <= timedelta(minutes=30):
        return [timedelta(minutes=15), timedelta(minutes=5)]

    return [timedelta(hours=1), timedelta(minutes=10)]


def _format_when(occ: Occ) -> str:
    if occ.all_day:
        return occ.start.strftime("%Y-%m-%d") + " (all-day)"
    return occ.start.strftime("%Y-%m-%d %I:%M %p ET")


def _reminder_key(occ: Occ, remind_at: datetime) -> str:
    return f"{occ.cal_name}|{occ.uid}|{occ.start.isoformat()}|{int(remind_at.timestamp())}"


def main() -> int:
    now = datetime.now(ET)
    horizon = now + timedelta(days=7)

    state = _load_state()
    sent = set(state.get("sent", []))

    reminders_to_send: List[str] = []

    # include occurrences starting slightly in the past so recurring expansion is stable
    occs = sorted(_iter_occurrences(now - timedelta(days=1), horizon), key=lambda o: o.start)

    for occ in occs:
        if occ.end <= now:
            continue

        offsets = _reminder_offsets(occ)

        if occ.all_day:
            remind_at = ET.localize(datetime(occ.start.year, occ.start.month, occ.start.day, 9, 0))
            if now <= remind_at <= now + timedelta(minutes=5):
                key = _reminder_key(occ, remind_at)
                if key not in sent:
                    reminders_to_send.append(
                        f"📅 Today ({occ.cal_name}): {occ.title} — {occ.start.strftime('%Y-%m-%d')} (all-day)"
                    )
                    sent.add(key)
            continue

        for off in offsets:
            remind_at = occ.start - off
            if now <= remind_at <= now + timedelta(minutes=5):
                key = _reminder_key(occ, remind_at)
                if key in sent:
                    continue

                if off >= timedelta(days=1):
                    label = "in 24h"
                elif off >= timedelta(hours=2):
                    label = "in 2h"
                elif off >= timedelta(hours=1):
                    label = "in 1h"
                else:
                    mins = int(off.total_seconds() // 60)
                    label = f"in {mins}m"

                reminders_to_send.append(
                    f"⏰ Reminder ({label}) [{occ.cal_name}]: {occ.title}\nWhen: {_format_when(occ)}"
                )
                sent.add(key)

    cutoff = now - timedelta(days=14)
    pruned: List[str] = []
    for k in sent:
        try:
            ts = int(k.split("|")[-1])
            if datetime.fromtimestamp(ts, ET) >= cutoff:
                pruned.append(k)
        except Exception:
            pruned.append(k)

    state["sent"] = sorted(set(pruned))
    state["updatedAt"] = now.isoformat()
    _save_state(state)

    if reminders_to_send:
        print("\n\n".join(reminders_to_send))

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/calendar/reminders.ts
#!/usr/bin/env npx tsx

import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { spawnSync } from "node:child_process";
import { readJsonFile, writeJsonFileAtomic } from "../lib/json-file.js";

type EventRow = { id: string; summary?: string; start?: string; end?: string; allDay?: boolean; calendar: string };

const STATE_PATH = path.join(os.homedir(), ".local/share/clawdbot/calendar-reminders-state.json");
const IMPORTANT_KEYWORDS = ["IMPORTANT", "TRAVEL", "FLIGHT", "DEADLINE", "INTERVIEW", "APPT"];

function now(): Date { return new Date(); }
function loadState(): any { return readJsonFile<any>(STATE_PATH) ?? {}; }

function listCalendars(): string[] {
  const p = spawnSync("gog", ["cal", "list", "--json"], { encoding: "utf8" });
  if (p.status !== 0 || !p.stdout.trim()) return ["Clawdbot-Calendar"];
  try {
    const js = JSON.parse(p.stdout);
    const arr = Array.isArray(js) ? js : js.calendars ?? js.items ?? [];
    const names = arr.map((x: any) => String(x.summary ?? x.name ?? x.id ?? "")).filter(Boolean);
    return names.length ? names : ["Clawdbot-Calendar"];
  } catch {
    return ["Clawdbot-Calendar"];
  }
}

function fetchEvents(cal: string, from = "yesterday", to: string): EventRow[] {
  const p = spawnSync("gog", ["cal", "list", cal, "--from", from, "--to", to, "--plain"], { encoding: "utf8" });
  if (p.status !== 0 || !p.stdout.trim()) return [];
  const lines = p.stdout.split("\n").filter((l) => l.trim());
  if (lines.length < 2) return [];
  const headers = lines[0].split("\t");
  return lines.slice(1).map((line) => {
    const vals = line.split("\t");
    const o: Record<string, string> = {};
    headers.forEach((h, i) => { o[h] = vals[i] ?? ""; });
    return {
      id: o.ID ?? "",
      summary: o.SUMMARY,
      start: o.START,
      end: o.END,
      allDay: (o.START ?? "").length === 10,
      calendar: cal,
    };
  }).filter((r) => r.id);
}

function offsets(ev: EventRow): number[] {
  const title = String(ev.summary ?? "").toUpperCase();
  const important = IMPORTANT_KEYWORDS.some((k) => title.includes(k));
  if (ev.allDay) return [0];
  if (important) return [24 * 60, 120, 30, 10];

  const s = new Date(ev.start ?? "");
  const e = new Date(ev.end ?? "");
  const durMin = Number.isNaN(s.getTime()) || Number.isNaN(e.getTime()) ? 60 : Math.max(1, (e.getTime() - s.getTime()) / 60000);
  if (durMin <= 30) return [15, 5];
  return [60, 10];
}

function reminderKey(ev: EventRow, ts: number): string {
  return `${ev.calendar}|${ev.id}|${new Date(ev.start ?? "").toISOString()}|${Math.floor(ts / 1000)}`;
}

function fmtWhen(ev: EventRow): string {
  const s = new Date(ev.start ?? "");
  if (ev.allDay) return `${(ev.start ?? "").slice(0, 10)} (all-day)`;
  return `${s.getFullYear()}-${String(s.getMonth() + 1).padStart(2, "0")}-${String(s.getDate()).padStart(2, "0")} ${s.toLocaleTimeString("en-US", { hour: "2-digit", minute: "2-digit", hour12: true })} ET`;
}

function main(): number {
  const n = now();
  const horizon = new Date(n.getTime() + 7 * 24 * 3600 * 1000);
  const to = horizon.toISOString().slice(0, 10);

  const state = loadState();
  const sent = new Set<string>(Array.isArray(state.sent) ? state.sent : []);
  const reminders: string[] = [];

  const calendars = listCalendars();
  let events: EventRow[] = [];
  for (const cal of calendars) events = events.concat(fetchEvents(cal, "yesterday", to));

  events.sort((a, b) => new Date(a.start ?? "").getTime() - new Date(b.start ?? "").getTime());

  for (const ev of events) {
    const start = new Date(ev.start ?? "");
    const end = new Date(ev.end ?? ev.start ?? "");
    if (Number.isNaN(start.getTime()) || end <= n) continue;

    if (ev.allDay) {
      const day = new Date(`${(ev.start ?? "").slice(0, 10)}T09:00:00`);
      if (n <= day && day <= new Date(n.getTime() + 5 * 60 * 1000)) {
        const key = reminderKey(ev, day.getTime());
        if (!sent.has(key)) {
          reminders.push(`📅 Today (${ev.calendar}): ${ev.summary ?? "(no title)"} — ${(ev.start ?? "").slice(0, 10)} (all-day)`);
          sent.add(key);
        }
      }
      continue;
    }

    for (const offMin of offsets(ev)) {
      const remindAt = new Date(start.getTime() - offMin * 60000);
      if (!(n <= remindAt && remindAt <= new Date(n.getTime() + 5 * 60 * 1000))) continue;
      const key = reminderKey(ev, remindAt.getTime());
      if (sent.has(key)) continue;

      const label = offMin >= 24 * 60 ? "in 24h" : offMin >= 120 ? "in 2h" : offMin >= 60 ? "in 1h" : `in ${offMin}m`;
      reminders.push(`⏰ Reminder (${label}) [${ev.calendar}]: ${ev.summary ?? "(no title)"}\nWhen: ${fmtWhen(ev)}`);
      sent.add(key);
    }
  }

  const cutoff = Date.now() - 14 * 24 * 3600 * 1000;
  const pruned = [...sent].filter((k) => {
    const ts = Number(k.split("|").at(-1));
    return Number.isNaN(ts) || ts * 1000 >= cutoff;
  });

  state.sent = [...new Set(pruned)].sort();
  state.updatedAt = n.toISOString();
  fs.mkdirSync(path.dirname(STATE_PATH), { recursive: true });
  writeJsonFileAtomic(STATE_PATH, state, 2);

  if (reminders.length) process.stdout.write(`${reminders.join("\n\n")}\n`);
  return 0;
}

process.exit(main());
